### PR TITLE
Fix the CBMC proof of prvTCPReturnPacket

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------

This change was required since the functioning has changed as compared to the IPv4 code base. There is a new concept of Interfaces and Endpoints which needs to be addressed in the proofs dealing with them.
This should technically be considered as the proofs being added rather than the proofs being changed since this is a new library (although based on IPv4 code base).

Test Steps
-----------

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c [hunk 1]
```diff
@@ -59,21 +60,93 @@ NetworkBufferDescriptor_t * pxDuplicateNetworkBufferWithDescriptor( NetworkBuffe
...
+ * implementation of this function in this proof. */
+BaseType_t NetworkInterfaceOutput( struct xNetworkInterface * pxDescriptor,
+                                   NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                   BaseType_t xReleaseAfterSend )
+{
+    BaseType_t xReturn;
+
+    /* Assert some basic conditions. */
+    __CPROVER_assert( pxDescriptor, "Descriptor cannot be NULL" );
+    __CPROVER_assert( pxNetworkBuffer, "NetworkBuffer cannot be NULL" );
```
Comment (markrtuttle) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c:74]: I think your coding standard is to explicitly compare to 0 or NULL.  But excellent job testing pointers in the stub!

File: test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c [hunk 2]
```diff
@@ -59,21 +60,93 @@ NetworkBufferDescriptor_t * pxDuplicateNetworkBufferWithDescriptor( NetworkBuffe
...
     {
-        pxNetworkBuffer->pucEthernetBuffer = safeMalloc( sizeof( TCPPacket_t ) );
-        __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer );
+        pxNetworkBuffer->pucEthernetBuffer = malloc( sizeof( TCPPacket_t ) );
+        __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer != NULL );
+
+        pxNetworkBuffer->pxEndPoint = &xLocalEndPoint;
+    }
+
+    if( ensure_memory_is_valid( pxSocket, sizeof( *pxSocket ) ) )
```
Comment (markrtuttle) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c:120]: This is the first time I've seen you use this function/macro, what is this?

**Review by markrtuttle — CHANGES_REQUESTED** [review]

File: test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c (hunk 2 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c:120]: This is a macro defined in memory_assignments.c which ensures that the memory is non NULL and writable
`#define ensure_memory_is_valid( px, length )    ( px != NULL ) && __CPROVER_w_ok( ( px ), length )`

File: test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c (hunk 1 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c:74]: Will do.

File: test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c (hunk 2 above)
Comment (AniruddhaKanhere) [reply test/cbmc/proofs/TCP/prvTCPReturnPacket/TCPReturnPacket_harness.c:120]: I will also add a clarifying comment

**Review by markrtuttle — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Pull request merged**